From 25c10f1246b58bef02557d622c53aed57c6363eb Mon Sep 17 00:00:00 2001
From: Felix Moessbauer <felix.moessbauer@siemens.com>
Date: Thu, 20 Aug 2026 10:57:43 +0200
Subject: [PATCH 1/1] rootless: bind-mount proc as fallback

Mounting the /proc from an inner namespace is not always possible due to
apparmor and seccomp profiles. Similar to mmdebstrap, we now rbind mount
proc instead in the inner namespace if mounting as procfs failed.

Note, that the inner mount anyways already runs in an unshared namespace
where a proper /proc was mounted. By that, the data visible to the inner
processes is the same.

Fixes: 4fedb1ae ("add support for fully rootless builds")
Signed-off-by: Felix Moessbauer <felix.moessbauer@siemens.com>
---
 meta/classes-global/base.bbclass | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meta/classes-global/base.bbclass b/meta/classes-global/base.bbclass
index e2fc77dc..89bca68f 100644
--- a/meta/classes-global/base.bbclass
+++ b/meta/classes-global/base.bbclass
@@ -400,7 +400,7 @@ def insert_isar_mounts(d, rootfs, mounts):
         lines.append('mount -t tmpfs none {}/dev/shm'.format(rootfs))
         lines.append('mount -o bind /dev/random {}/dev/random'.format(rootfs))
         lines.append('mount -o bind /dev/urandom {}/dev/urandom'.format(rootfs))
-        lines.append('mount -t proc none {}/proc'.format(rootfs))
+        lines.append('mount -t proc -o ro none {0}/proc || mount -o rbind /proc {0}/proc'.format(rootfs))
         lines.append('ln -sfn /proc/self/fd {}/dev/fd'.format(rootfs))
         lines.append('ln -sfn /proc/self/fd/0 {}/dev/stdin'.format(rootfs))
         lines.append('ln -sfn /proc/self/fd/1 {}/dev/stdout'.format(rootfs))
-- 
2.55.0

